timemachine/fe/mle.py

```diff
@@ -212,3 +214,103 @@ def infer_node_vals_and_errs(
     dg_err = bootstrap_estimates.std(0)
 
     return dg, dg_err
+
+
+def infer_node_vals_and_errs_networkx(
+    nx_graph: nx.Graph,
+    edge_diff_prop: str,
+    edge_stddev_prop: str,
+    node_val_prop: str,
+    node_stddev_prop: str,
+    ref_nodes: Sequence[Any],
+    ref_node_val_prop: str,
+    ref_node_stddev_prop: str,
+    n_bootstrap: int = 100,
+    seed: int = 0,
+) -> nx.Graph:
+    """Version of :py:func:`timemachine.fe.mle.infer_node_vals_and_errs` that accepts a networkx graph.
+
+    Parameters
+    ----------
+    nx_graph: networkx.Graph
+        Networkx graph
+    edge_diff_prop: str
+        Edge property to use for differences
+    edge_stddev_prop: str
+        Edge property to use for standard deviations
+    node_val_prop: str
+        Node property in which to record inferred absolute values
+    node_stddev_prop: str
+        Node property in which to record inferred standard deviations
+    ref_nodes: sequence
+        Reference nodes (must have properties `ref_node_val_prop` and `ref_node_stddev_prop`)
+    ref_node_val_prop: str
+        Node property to use for reference values. Must be defined on reference nodes.
+    ref_node_stddev_prop: str
+        Node property to use for reference standard deviations. If undefined in a reference node, assumed to be zero.
+    n_bootstrap, seed:
+        See documentation for :py:func:`timemachine.fe.mle.infer_node_vals_and_errs`
+
+    Returns
+    -------
+    networkx.Graph
+        Graph where all nodes have been labeled with the inferred value of `node_val_prop` and `node_stddev_prop`.
+    """
+
+    def get_valid_subgraph(g: nx.Graph) -> nx.Graph:
+        "Remove edges with missing edge_diff_prop or edge_stddev_prop; then, remove any isolated nodes"
+        sg = g.copy()
+
+        missing_edges = [e for e, v in sg.edges.items() if not v.get(edge_diff_prop) or not v.get(edge_stddev_prop)]
+        sg.remove_edges_from(missing_edges)
+
+        isolated_nodes = list(nx.isolates(sg))
+        sg.remove_nodes_from(isolated_nodes)
```

Comment:
Uncertainty: Is "nodes with zero degree (`nx.isolates`) the intended concept?

(May want to remove nodes that aren't in the largest connected component, even if they have degree > 0)

Reply:
Handling missing edges seems more subtle -- would it be reasonable to split this into a separate PR?

I'm uncertain about the right behavior when edges are missing:
* keep all nodes with degree > 0 vs. keep largest connected component vs. keep component that contains all ref_nodes
* throw error vs. not if some ref_nodes are in different disconnected components
* return copy of full graph where some nodes may not have mle attributes vs. return copy of subgraph where all nodes have mle attributes...

For now, perhaps this function can assert as a pre-condition that all edges in the graph have valid values for `edge_diff_prop`?

Reply:
> Uncertainty: Is "nodes with degree zero" (nx.isolates) the intended concept?

It was intended; this is to prevent the assertion https://github.com/proteneer/timemachine/blob/9ac44c207b21750f15c5fc3b8156e5516e131438/timemachine/fe/mle.py#L51 from failing in the case where removing edges results in some nodes that don't appear in the edge list. It seemed more appropriate to (eventually) handle the case of multiple connected components in `infer_node_vals_and_errs`.

Reply:
https://github.com/proteneer/timemachine/pull/987#discussion_r1154830927

Ahh, sorry, I forgot most of these are already ruled out by https://github.com/proteneer/timemachine/blob/9cb5204566c06461653dc83aec031d56ddc47d1d/timemachine/fe/mle.py#L52-L54  -- thanks for the reminder @mcwitt .

I think this looks good.

Reply:
(The "disconnected components but node degree > 0" case is already ruled out -- the current PR only needs to handle the new case now covered by `nx.isolates`, as implemented.)